python/lsst/cp/pipe/ptc.py

```diff
@@ -369,147 +362,218 @@ def runDataRef(self, dataRef, visitPairs):
+
+        if self.config.ptcFitType in ["ASTIERFULL", ]:
+            # Calculate covariances and fit them, including the PTC, to Astier+19 full model (Eq. 20)
+            datasetPtc = self.fitCovariancesAstier(datasetPtc, covariancesWithTags)
+        elif self.config.ptcFitType in ["ASTIERAPPROXIMATION", "POLYNOMIAL"]:
+            # Fit the PTC to a polynomial or to Astier+19 approximation (Eq. 16)
+            # Fill up PhotonTransferCurveDataset object.
+            datasetPtc = self.fitPtc(datasetPtc, self.config.ptcFitType)
+
+        # Fit a poynomial to calculate non-linearity and persist linearizer.
+        if self.config.doCreateLinearizer:
+            numberAmps = len(amps)
+            numberAduValues = self.config.maxAduForLookupTableLinearizer
+            lookupTableArray = np.zeros((numberAmps, numberAduValues), dtype=np.float32)
+
+            # Fit (non)linearity of signal vs time curve.
+            # Fill up PhotonTransferCurveDataset object.
+            # Fill up array for LUT linearizer (tableArray).
+            # Produce coefficients for Polynomial ans Squared linearizers.
+            # Build linearizer objects.
+            datasetPtc, linsArray = self.fitNonLinearityAndBuildLinearizers(datasetPtc, detector,
+                                                                            tableArray=lookupTableArray,
+                                                                            log=self.log)
+
+            butler = dataRef.getButler()
+            self.log.info("Writing linearizers: \n "
+                          "lookup table (linear component of polynomial fit), \n "
+                          "polynomial (coefficients for a polynomial correction), \n "
+                          "and squared linearizer (quadratic coefficient from polynomial)")
+
+            for (linearizer, dataType) in linsArray:
+                detName = detector.getName()
+                now = datetime.datetime.utcnow()
+                calibDate = now.strftime("%Y-%m-%d")
+
+                butler.put(linearizer, datasetType=dataType, dataId={'detector': detNum,
+                           'detectorName': detName, 'calibDate': calibDate})
+
+        self.log.info(f"Writing PTC data to {dataRef.getUri(write=True)}")
+        dataRef.put(datasetPtc, datasetType="photonTransferCurveDataset")
 
-                dataset.rawExpTimes[ampName].append(expTime)
-                dataset.rawMeans[ampName].append(mu)
-                dataset.rawVars[ampName].append(varDiff)
-                dataset.inputVisitPairs[ampName].append((v1, v2))
+        return pipeBase.Struct(exitStatus=0)
 
-        numberAmps = len(detector.getAmplifiers())
-        numberAduValues = self.config.maxAduForLookupTableLinearizer
-        lookupTableArray = np.zeros((numberAmps, numberAduValues), dtype=np.float32)
+    def fitNonLinearityAndBuildLinearizers(self, dataset, detector, tableArray=None, log=None):
```

Comment:
Can this be moved down to where fitNonLinearity is? Grouping the PTC methods separate from the Linearizer methods makes it much easier to trace through the code without big jumps.  It's possible your editor makes it easy to find related methods.